### ISSUE TITLE
Change masked wrapper to use mask_fn instead of mask

### ISSUE THREAD

**Issue by n2cholas** [deepmind/optax#57]
Currently, `optax.masked` accepts a boolean mask that has the same structure as (or is a prefix of) the parameters pytree. This breaks the pattern of only requiring the parameters during `init` and not before. This issue is to discuss the possibility of changing the `mask` argument to `mask_fn`, which would be a function that takes a parameter pytree as input and returns a mask that has the same/prefix structure as the params. 

One clear advantage of the `mask_fn` approach is users can define an optimizer independent of the model (as is the case with every other transformation in optax).

The current use pattern is still possible by passing in `mask=lambda _: mask` for a premade `mask` if desired.

This was originally proposed by @jheek in the [New Optimizer API](https://github.com/google/flax/pull/1011) for Flax discussion.

cc @mtthss @andsteing

**Comment by mtthss** [comment]
Sounds good,

Lets make `mask` be either a `PyTree` or a `Callable`

**Comment by n2cholas** [comment]
Great! One slight issue is the `inject_hyperparams` wrapper (#48) has no way of telling whether a callable is a schedule or a fixed hyperparameter, which would make it not compose properly with this proposed `masked`.

An immediate solution that comes to mind is to add an `ignore` argument to `inject_hyperparams`, which accepts a list of hyperparameters to ignore for scheduling:

```python
inject_hyperparams(custom_optimizer, ignore=['mask'])(
    learning_rate=schedule(...), mask=mask_fn)
```

I don't think adding an optional argument is too bad for this edge case, but perhaps a more elegant/robust solution would be better. I'll think about this more before creating a PR.
